### PR TITLE
Password encryption design

### PR DESCRIPTION
**Pull request by jackhaibo**
Password  encryption design. @xing-yang @wisererik @leonwanghui Please take a review at this PR, thanks!

### CODE REVIEW

File: specs/capri/Passwor_Encryption_Design.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+**Author(s)**: [Jack Li](https://github.com/lasshpanzer)
+
+## Summary
+
+This document is a design proposal to provide a password security mechanism for OpenSDS.
+
+## Motivation
+
+Currently, the password used by other components such as cinder, kubernetes, MultiCloud to interact with OpenSDS is not encrypted, which causes the user’s information to be easily leaked.
```
Comment (leonwanghui) [inline specs/capri/Passwor_Encryption_Design.md]: s/to be/being

Comment (jackhaibo) [reply specs/capri/Passwor_Encryption_Design.md]: Already changed. Please review again.

File: specs/capri/Passwor_Encryption_Design.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+2.Provide OpenSDS default AES password encryption tool, so that user can use this tool to get cipher text and shell scripts can call it for automatic encryption during the deployment process. 
+
+AES encryption tool user guide：
+build/out/bin/pwdEncrypter password
+
+3.Provide a unified code framework for encryption and decryption that enables 
+user to implement their own encryption tool.
+
+type PwdEncrypter interface {
```
Comment (leonwanghui) [inline specs/capri/Passwor_Encryption_Design.md]: I would suggest to change its format like this:
```go
type PwdEncrypter interface {
	Encrypter(password string) (string, error)
	Decrypter(code string) (string, error)
}
```

File: specs/capri/Passwor_Encryption_Design.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	Encrypter(password string) (string, error)
+	Decrypter(code string) (string, error)
+}
+
+### Data model impact
+
+Add two attributes PwdEncrypter and EnableEncrypted in AuthOptions struct.
+type AuthOptions struct {
```
Comment (leonwanghui) [inline specs/capri/Passwor_Encryption_Design.md]: Same here

File: specs/capri/Passwor_Encryption_Design.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+None
+
+### Performance impact
+
+None
+
+### Other deployer impact
+
+When multi-cloud, dorado, cinder, kubernetes, keystone docking the OpenSDS, user needs to change the value of enable_encrypted and pwd_encrypter in the configuration file to decide whether to encrypt the password and choose which encryption tool to use or keep the default values.
```
Comment (leonwanghui) [inline specs/capri/Passwor_Encryption_Design.md]: s/default values/default one

Comment (leonwanghui) [inline specs/capri/Passwor_Encryption_Design.md]: s/docking/connect with

File: specs/capri/Passwor_Encryption_Design.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+### Developer impact
+
+Developers can implement their own encryption and decryption functions in the OpenSDS encryption framework according to their own needs.
```
Comment (leonwanghui) [inline specs/capri/Passwor_Encryption_Design.md]: s/according to their own needs/based on their requirements

**Comment by xing-yang** [comment]
Please move to https://github.com/opensds/design-specs/tree/master/specs/capri

**Comment by jackhaibo** [comment]
> Could you add the introduction of which parts of opensds should be encrypted? The reason for doing this is to give a better view of what parts should be taken care of.
All issues fixed, please review again.

File: specs/capri/Passwor_Encryption_Design.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+Currently, the password used by other components such as cinder, kubernetes, MultiCloud to interact with OpenSDS is not encrypted, which causes the user’s information being easily leaked.
+
+### Goals
+
+The password shall be encrypted when other components use it to interact with OpenSDS.
+
+1.A default encryption tool will be provided for users to obtain the cipher text.
+
+2.Which encryption tool is selected and whether the password is encrypted can be configurable. By default, the encryption tool is AES provided by the community, and the password is not encrypted.
+
```
Comment (xing-yang) [inline specs/capri/Passwor_Encryption_Design.md]: s/AES provided by the community/AES which is an open source tool

File: specs/capri/Passwor_Encryption_Design.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+### Non-Goals
+
+None
```
Comment (xing-yang) [inline specs/capri/Passwor_Encryption_Design.md]: Remove the "Non-Goals" section.

File: specs/capri/Passwor_Encryption_Design.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+## Design Details
+
+1.Add options to the OpenSDS configuration file that enable user to choose the password encrypter tool they want and choose whether to encrypt the password. The encryption tools can be implemented by themselves. These two options added to the opensds.conf are as follows:
+
```
Comment (xing-yang) [inline specs/capri/Passwor_Encryption_Design.md]: s/can be implemented by themselves/can be implemented by the user

File: specs/capri/Passwor_Encryption_Design.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    }
+    opt.Password = pwdCiphertext
+
+2.Provide OpenSDS default AES password encryption tool, so that user can use this tool to get cipher text and shell scripts can call it for automatic encryption during the deployment process. 
+
+AES encryption tool user guide：
+build/out/bin/pwdEncrypter password
+
+3.Provide a unified code framework for encryption and decryption that enables 
+user to implement their own encryption tool.
```
Comment (xing-yang) [inline specs/capri/Passwor_Encryption_Design.md]: Move line 56 up.  Don't need to start a new line.

File: specs/capri/Passwor_Encryption_Design.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+Increased user information security.
+
+### Other end user impact
+
+None
+
+### Performance impact
+
+Note:These parts of hotpot should be considered encrypted:
+Southward volume drivers or other drivers: multi-cloud, dorado, fusionstorage, cinder. These parts use user name and password to connect with hotpot.
```
Comment (xing-yang) [inline specs/capri/Passwor_Encryption_Design.md]: s/Southward/Southbound

File: specs/capri/Passwor_Encryption_Design.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+Northward plugin:csi. These parts use user name and password to get token from keystone.
```
Comment (xing-yang) [inline specs/capri/Passwor_Encryption_Design.md]: s/Northward/Northbound

File: specs/capri/Passwor_Encryption_Design.md [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+**Author(s)**: [Jack Li](https://github.com/lasshpanzer)
+
+## Summary
+
+This document is a design proposal to provide a password security mechanism for OpenSDS.
+
+## Motivation
+
+Currently, the password used by other components such as cinder, kubernetes, MultiCloud to interact with OpenSDS is not encrypted, which causes the user’s information being easily leaked.
+
```
[the remark this answers is not in the document]
Comment (jackhaibo) [reply specs/capri/Passwor_Encryption_Design.md]: All issues has been fixed. Please review again. Thanks

File: specs/capri/Passwor_Encryption_Design.md [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+## Motivation
+
+Currently, the password used by other components such as cinder, kubernetes, MultiCloud to interact with OpenSDS is not encrypted, which means the user’s information can be leaked easily.
+
+### Goals
+
+The password shall be encrypted when other components use it to interact with OpenSDS.
+
+1.A default encryption tool will be provided for users to obtain the cipher text.
```
Comment (xing-yang) [inline specs/capri/Passwor_Encryption_Design.md]: Need a space betwee "1." and "A default...".

File: specs/capri/Passwor_Encryption_Design.md [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+
+2.Which encryption tool is selected and whether the password is encrypted can be configurable. By default, the encryption tool is AES which is an open source tool, and the password is not encrypted.
```
Comment (xing-yang) [inline specs/capri/Passwor_Encryption_Design.md]: Need a space after "2.".

File: specs/capri/Passwor_Encryption_Design.md [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+
+
+## Design Details
+
+1.Add options to the OpenSDS configuration file that enable user to choose the password encrypter tool they want and choose whether to encrypt the password. The encryption tools can be implemented by the user. These two options added to the opensds.conf are as follows:
```
Comment (xing-yang) [inline specs/capri/Passwor_Encryption_Design.md]: Need a space after "1."

File: specs/capri/Passwor_Encryption_Design.md [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+
+// Whether to encrypt the password. If enabled, the value of the password must be cipher text.
+enable_encrypted = False
+// Encryption and decryption tool. Default value is aes. The decryption tool can only decrypt the corresponding ciphertext.
+pwd_encrypter = aes
+
```
Comment (xing-yang) [inline specs/capri/Passwor_Encryption_Design.md]: Add "\`\`\`" on the line before 26.  Add "\`\`\`" on the line after "pwd_encrypter":

\`\`\`
// Whether to encrypt the password. If enabled, the value of the password must be cipher text.
enable_encrypted = False
// Encryption and decryption tool. Default value is aes. The decryption tool can only decrypt the corresponding ciphertext.
pwd_encrypter = aes
\`\`\`

File: specs/capri/Passwor_Encryption_Design.md [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+        pwdTool := pwd.NewPwdEncrypter(opt.PwdEncrypter)
+        // Decrypt the password and obtain the original password.
+        password, err := pwdTool.Decrypter(pwdCiphertext)
+        if err != nil {
+	        return nil, err
+        }
+        pwdCiphertext = password
+    }
+    opt.Password = pwdCiphertext
+
```
Comment (xing-yang) [inline specs/capri/Passwor_Encryption_Design.md]: Add \`\`\` before the code and after the code.

File: specs/capri/Passwor_Encryption_Design.md [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+2.Provide OpenSDS default AES password encryption tool, so that user can use this tool to get cipher text and shell scripts can call it for automatic encryption during the deployment process.
```
Comment (xing-yang) [inline specs/capri/Passwor_Encryption_Design.md]: Space after "2."

File: specs/capri/Passwor_Encryption_Design.md [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+        pwdCiphertext = password
+    }
+    opt.Password = pwdCiphertext
+
+2.Provide OpenSDS default AES password encryption tool, so that user can use this tool to get cipher text and shell scripts can call it for automatic encryption during the deployment process. 
+
+AES encryption tool user guide：
+build/out/bin/pwdEncrypter password
+
+3.Provide a unified code framework for encryption and decryption that enables user to implement their own encryption tool.
```
Comment (xing-yang) [inline specs/capri/Passwor_Encryption_Design.md]: Space after "3."

File: specs/capri/Passwor_Encryption_Design.md [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+
+type PwdEncrypter interface {
+	Encrypter(password string) (string, error)
+	Decrypter(code string) (string, error)
+}
```
Comment (xing-yang) [inline specs/capri/Passwor_Encryption_Design.md]: Add ``` before the code and after the code.

File: specs/capri/Passwor_Encryption_Design.md [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+type AuthOptions struct {
+	Strategy        string `yaml:"Strategy"`
+	AuthUrl         string `yaml:"AuthUrl,omitempty"`
+	DomainName      string `yaml:"DomainName,omitempty"`
+	UserName        string `yaml:"UserName,omitempty"`
+	Password        string `yaml:"Password,omitempty"`
+	PwdEncrypter    string `yaml:"PwdEncrypter,omitempty"`
+	EnableEncrypted bool   `yaml:"EnableEncrypted,omitempty"`
+	TenantName      string `yaml:"TenantName,omitempty"`
+}
```
Comment (xing-yang) [inline specs/capri/Passwor_Encryption_Design.md]: Add ``` before the code and after the code.

File: specs/capri/Passwor_Encryption_Design.md [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+## Use Cases
+
+Before deployment, user change EnableEncrypted to true and keep PwdEncrypter option as default in the configuration file, then use aes encryption tool get cipher text and write it to the password option.
+
+## Implementation
+
+1.Implement the encryption tool framework.
+2.Implement AES encryption and decryption functions.
+3.Implement AES encryption CLI tool.
+4.Modify multi-cloud, dorado, cinder, kubernetes and keystone configuration file, AuthOptions struct and corresponding code.
```
Comment (xing-yang) [inline specs/capri/Passwor_Encryption_Design.md]: Need a space after "1." etc.

Comment (jackhaibo) [reply specs/capri/Passwor_Encryption_Design.md]: All issues were fixed. Please review again, thanks. @xing-yang

**Comment by xing-yang** [comment]
lgtm

**Pull request merged**